Sources/AsyncStore/Effect.swift

```diff
@@ -32,6 +32,20 @@ public extension AsyncStore.Effect {
         .task(operation: operation, id: id)
     }
     
+    static func debounce(
+        operation: @escaping () async throws -> Self,
+        id: AnyHashable,
+        delay: TimeInterval
+    ) -> Self {
+        .task(
+            operation: {
+                try await Task.trySleep(for: delay)
```

Comment:
I see you merged already, but this is actually a delay (or really a delayEach) rather than a debounce unless you have forced the upstream to have `bufferingOldest(1)` as a buffering policy.  debounce needs to drop any additional arrivals during the debounce period and its tricky to get right.

Reply:
I see what you're saying.  But here we are not debouncing the effect, we are debouncing the task itself.  This is a pattern in the Async Store that can be done without this "convenience" effect, but this makes the pattern more apparent.   

The pattern being:
* identify a task
* delay the task's execution and cancel it if another task is scheduled with the same ID

What we're really doing here, is forcing the consumer to identify the task, and for convenience, adding the sleep before task execution.  This prevents task collisions.

Reply:
@rvsrvs I thought that the debounce doesn't drop the incoming calls, but continues to reset the interval with new input.   If you use receiving text input as an example, and you only want to perform some action after half a second, using `func search(string: String)` as an example. 

search("a") // receive input and start timer
search("ab") IF half a second has passed, hold this input and start timer ELSE reset existing timer with this new input

Is debounce the right term for this? 

I initially thought that the way Wendell used sleep here, this functioned as a delay or delayEach as you say, but it's the `id` that comes in that allows you to overwrite the call being made if it happens during that sleep period.  Effectively any existing tasks with that ID that are received during the sleep will be canceled

Reply:
```
Publishers.Debounce
A publisher that publishes elements only after a specified time interval elapses between events.
```
In Combine terminology, the upstream publisher can publish at any time (i.e. is generally a Subject).  Every time a new value is accepted and passed downstream, a timer gets set and anything new that is sent before the Timer goes off is discarded.  Try the following in a playground:
```
import Combine
import Dispatch

let subject = PassthroughSubject<Int, Never>()

let c = subject
    .debounce(for: 1.0, scheduler: DispatchQueue.main)
    .sink { value in
        print(value)
    }

subject.send(1)
subject.send(2)
subject.send(3)
subject.send(4)
subject.send(5)
sleep(2)
subject.send(6)
```
The output will be 1 and 6.  The difference between this and throttle is that throttle has a constantly running timer that does not get reset on successful sends.  What you are describing is actually a zip of the input with a timer.  More like this:
```
let t = Timer.publish(every: 2.0, on: RunLoop.main, in: .common).autoconnect()
let c1 = "abcdefg".publisher
    .zip(t)
    .sink { value in
        print(value)
    }
```
That will produce output like:
```
("a", 2022-07-11 00:55:01 +0000)
("b", 2022-07-11 00:55:03 +0000)
("c", 2022-07-11 00:55:05 +0000)
("d", 2022-07-11 00:55:07 +0000)
("e", 2022-07-11 00:55:09 +0000)
("f", 2022-07-11 00:55:11 +0000)
("g", 2022-07-11 00:55:13 +0000)
```
Which I think is what you are wanting..

Reply:
The temporal operators in FRP are among the hardest to get right.  I've spent a lot of time puzzling over this.

Reply:
This is essentially what the unit test for the “debounce” effect does.  Perhaps the term is not accurate for the problem we’re solving but I think it is.

Reply:
@rvsrvs I think your example is proving what I am saying about debounce.  Your initial comment ```debounce needs to drop any additional arrivals during the debounce period and its tricky to get right.``` 

I think your last comments agrees with what I was saying that it doesn't drop additional arrivals during the debounce, but receives and resets the timer each time.

If you remove all of those sleeps in your code, the only output will be `6` because it will ignore all the previous calls, not all the calls after.  The calls passing 1 through 5 are all ignored because the one calling 6 was the last call and then no input was received in that threshold so it sends.  

The idea of debounce using the first input and then ignoring all the calls after for a specific time wouldn't make as much sense programmatically because you could send a bunch of calls into the void.  With a proper debounce, you only ignore if you get a new input in X interval.

Reply:
@rvsrvs just for clarity, I'm not trying to be argumentative, just trying to make sure I Understood your comment and that I had the correct understanding of debounce and was trying to reconcile that with what you were saying and the example you provided.

Reply:
i did misstate in my original comment   debounce requires bufferingNewest(1) not bufferingOldest(1).  argh.

debounce and throttle require you to drop values if they arrive too quickly. that's definitional and comes from electrical engineering. Combine is doing something exactly analagous to the behavior of mechanical debouncers.  if you don't drop values based on interarrival times you don't have the debounce or throttle operators correct. 

what you have here is a different operator. i was commenting on the name.  this operator may be what you want but it is not debounce and will not replicate the behavior of debounce operators in any frp system.

bigger picture: what you are wrestling with here is the difference between a pull stream and a push stream.  combine is push, the way you are using asyncstream is pull.  to get actual debounce behavior, you have to do push.  in this case, you'd need to create an additional asyncstream in line but in a different task and give it bufferingNewest(1) semantics.  but then you are off to the Continuation races and thats a much bigger task (pun intended)